Feat/proxy v0.1

m4ntis opened the pull request:
#### Hey all,

As some of you maintainers probably know by now, for the last couple of months me and 2 colleagues of mine (@kubistika and @speidy) have been working on implementing an RDP proxy using the library. This is a PR to introduce what we've been working on up till now, containing a working base version of the proxy.

The changes and number of commits are quite vast, so I'll start with the motivation, why we chose to implement it in this way and why the *h* you might want all of this massive PR. Later on I'll get to what is implemented here, how to use it, future goals and where we want to get to with this proxy.

## Motivation
The proxy came about for a number of reasons, mostly security related:
- There is no real good RDP proxy out there, enabling you to finely select the specific features and channels you want to allow, filter or right out disable for all connections passing through. Lets say you have some isolated network where you want to minimize all traffic going in and out while still enabling easy access via RDP, you probably don't want any channels passing through but the bare minimum, say GFX from the server and basic input events from the client.
- There is no proxy that lets you monitor and analyze RDP traffic going through it. We might want to log certain events or look for anomalies, such as suspicious amounts of certain events or specific channel messages.
- We wanted a proxy that will let you capture and record all traffic passing through. You might use this feature by placing the proxy as the only gateway to some sensitive server such as a DC, being able to view all passed connections filtered by time and username.
- We wanted to have a proxy that has an API for extra authentication, letting us implement a 2FA towards the proxy and disallowing connections to the target server without authentication.

## Why we chose FreeRDP
So the reasoning for choosing to implement the proxy as a part of the FreeRDP library is quite obvious. FreeRDP is by far the most maintained and up to date OS implementation of the protocol, and with the server side implementations only getting bigger, it only seemed fitting. As we **do** parse the messages going through the proxy (the whole point of the proxy), in order to support updated versions of the protocol we need to use an updating implementation of RDP. There is a tiny, basic proxy MITM in [pyrdp](https://github.com/GoSecure/pyrdp) but it's too small, not extensive and with way lower chances of being properly maintained in the long run.

## Benefits to the library
Well first of all, we hope that the proxy in itself is a useful thing to have on it's own. But further than that, as you've seen in the last month or so, me and @kubistika already issued a few small PRs that fixed some of the server side implementation of the protocol. This is just the beginning, and this PR already has a few other fixes to GFX parsing bundled inside it (I'll elaborate below). The proxy tests both client and server side implementations of the protocol without having to implement a client nor server, as you can easily use mstsc and Microsoft's RDP TS server going talking via the proxy to quickly find any parsing errors.

I'm sure we'll continue to squish bugs as we continue to other parts of the protocol :wink:.

Additionally, the proxy is an incentive for us to implement the server side of a few other channels that aren't implemented yet, as we want to parse and proxy them too (I am currently working on a RAIL server implementation, might be a few other coming after that).

## So what do we have here
This is the first usable version of the proxy, so most of our goals aren't yet fully implemented (some not at all). What we do have is a proxy that can parse and proxy connections, passing graphics messages and keyboard / mouse events. GFX and bitmap updates are fully working, GDI still has some issues but I think that it's obvious why GDI is a low priority right now.

**Basic list of features**:
- GFX, bitmap updates and input events parsing and proxying.
- Settings sync: A client connects to and negociates with the proxy, the proxy takes the negociated settings and connects with them to the server, and then renegociates with the client using the final settings.
- GFX capabilities sync: The client and server capabilities are properly synced.
- Basic configuration (example below). Not all features in the config file are fully implemented yet, but soon to be.

There are a few changes to the library itself too:
#### Bugfixes:
1. In `client/rdpgfx_main.c:rdpgfx_recv_caps_confirm_pdu()`, the caps data length was incorrectly read into a local variable.
2. In `client/rdpgfx_main.c:rdpgfx_recv_map_surface_to_scaled_window_pdu()` and `scaled_output_pdu`, the stream length magic number was incorrect, as the methods were probably copy+pasted from `recv_map_surface_to_window/output` without changing the length.
3. Same issue as the one above in the server side, in `server/rdpgfx_main.c:rdpgfx_send_map_surface_to_scaled_window_pdu`, stream length is incorrect.
4. Same again with `scaled_output_pdu` on the server side, but on this PDU the values in the struct were named wrongly by the spec, `targetX` and `targetY` instead of `targetWidth` and `targetHeight`.

#### Changes to API:
1. In order to properly be able to forward GFX frame acking and to sync capabilities between the client and server, we needed to have a way to have control over sending FrameAck and CapsConfirm messages, so we exposed these methods as part of the `RdpgfxClientContext`'s API.
2. Additionally, we needed some way of disabling the client's automatic frame acking / caps confirmation, without hindering default functionality. We added and optional `OnOpen` callback with flags that control whether to `do_frame_acks` and `do_caps_confirm`. These values default to `TRUE`.

## Usage
To compile with cmake, the `-DWITH_SERVER=on` flag will compile the proxy. Running it is as simple as `./pfreerdp` (there are no CLI options yet). There is a `config.ini` present in `server/proxy` that currently has to be in the working dir of the proxy. Lets go over the config:

```ini
[Server]
Host = "0.0.0.0"
Port = 3389
LocalOnly = 0

[Target]
UseLoadBalanceInfo = 1
Host = "Target server"
Port = 3389

[Input]
Mouse = 1
Keyboard = 1

[Graphics]
GFX = 1
BitmapUpdate = 1

[Security]
NlaSupport = 0
TlsSupport = 0
RdpSupport = 0

[Channels]
WhitelistMode = 0
AllowedChannels = "cliprdr,Microsoft::Windows::RDS::Video::Control"
DeniedChannels = "Microsoft::Windows::RDS::Geometry"
```
All sections but the `[Target]` and `[Channels]` are quite obvious (host config and allowing / disallowing RDP features).

The `[Target]` config option regards the method of conveying which server shall the client connecting to the proxy be forwarded to. The simple option is configuring a `Host` and `Port` value and setting `UseLoadBalanceInfo` to `0`. This fixes all connections to a single configured destination. The second method is to dynamically detect the target for each incoming connection. The way we figured out is to use the routing token to send that info (this is one of the things we would like to hear your opinion on). This allows the proxy to receive the target server info from any normal rdp client. For mstsc, you could use the `loadbalanceinfo:s:` option in an rdp file, and in any of the FreeRDP clients you could use the `/load-balance-info:` flag. The format of the routing token is `Cookie: msts=<token>`, so as an example, using with xfreerdp would look like this:

`xfreerdp /v:<proxy addr> /load-balance-info:"Cookie: msts=<destination server>"`
 
The `[Channels]` section is not yet implemented, but expresses the two modes we want the proxy to be able to work in, blacklist and whitelist. The first would allow all channels excluding the ones configured in `DeniedChannels`. The latter would allow only the channels configured in `AllowedChannels`. Currently, only the channels explicitly supported by the proxy are allowed. In order to support a blacklist mode, we would have to find a way of proxing channels without inspecting the data. We thought about implementing a "dummy" channel that just exposes receiving and sending data (we'd love hearing suggestions about this too).

### Small caveat
NLA isn't supported yet, but will be very soon.

## Looking forward
I'll keep this short, but here's a short summary of the features we want to implement:

1. A hooking type of API for events in the connection, probably loading "so" files that register hooks the proxy exposes.
2. Implementing more channels.
3. Implementing a "blacklist" channel mode (explained just above)
4. Adding some sort of API for a third party authentication logic.
5. Connection recording functionality.

This list can change / grow, we're willing to hear suggestions for features / how to implement them.

Please note that this is just the first initial PR for this proxy. We are and will be actively developing this proxy, adding support for more channels and features. We wanted to do this PR and merge early so that we could get a lot of feedback and suggestions from the maintainers, and so that you follow and review our work.

Thanks for taking the time to read and review this big PR :)

akallabeth committed:
Added a proper synthesizer for FileGroupDescriptorW to text/uri-list

The file clipboard delegate needs a base URI to operate on for
systems that are not WIN32. Added that to the context and abort
conversion, if that is not set. (currently not fully implemented)

m4ntis committed:
server: Add proxy dir with barebones server

m4ntis committed:
sever/proxy: Remove licensing

kubistika committed:
server/proxy: Add client files

m4ntis committed:
server/proxy: rm binary

m4ntis committed:
server/proxy: Formatting

kubistika committed:
server/proxy: Fixed includes and added basic client creation functionality

m4ntis committed:
server/proxy: Remove licensing and fix ifndef

m4ntis committed:
proxy/server: Fix cmake indentation

m4ntis committed:
server/proxy: Fix licensing

kubistika committed:
server/proxy: Forward connection on peer_post_connect

kubistika committed:
server/proxy: Fix function signature

kubistika committed:
server/proxy: Changed function signature of proxy_client_start

kubistika committed:
server/proxy: Now peer_post_connect calls proxy_client_start in a new thread

m4ntis committed:
pfreerdp.c: Clean up useless comments and logs

m4ntis committed:
server/proxy: Fix license

m4ntis committed:
server/proxy: Remove all non-connection related data from proxy_context

m4ntis committed:
server/proxy: Move Log Tag definition to pf_log.h

m4ntis committed:
server/proxy: Move context definition to pf_context

m4ntis committed:
server/proxy: Delete pfreerdp.h

m4ntis committed:
pfreerdp.c: Move context callbacks to pf_context.c

m4ntis committed:
server/proxy: Update CMakeLists.txt

m4ntis committed:
pf_channels: Use new proxy context API

m4ntis committed:
pf_client: Move context to pf_context

m4ntis committed:
pf_client.c: Remove unnessecary event handling

m4ntis committed:
server/proxy: Formatting

m4ntis committed:
proxy/server: Move server logic to pf_server.c

kubistika committed:
server/proxy: Handle client disconnection

m4ntis committed:
Merge stash

m4ntis committed:
pf_server.c: Open GFX Connection to client

speidy committed:
server: CMakeLists: build proxy along with other servers

speidy committed:
server: proxy: get target server from rdpNego->RoutingToken

Iv'e omitted a check from  which im not sure is right. Should check in docs

kubistika committed:
server/proxy: Handle remote server -> client disconnection

kubistika committed:
server/proxy: Move common function  to pf_common.c

kubistika committed:
server/proxy: Move common function  to pf_common.c

m4ntis committed:
rdpgfx.h: Add reference to freerdp.h for rdpContext

m4ntis committed:
pf_channels: Pipe GFX on channel connection

m4ntis committed:
server/proxy: Add pf_rdpgfx for proxy gfx callbacks

m4ntis committed:
pf_client: Declare dynvc and gfx capabilities on connection

kubistika committed:
server/proxy: Add graphics callbacks

kubistika committed:
server/proxy: Add graphics callbacks

m4ntis committed:
pf_server: Listen to channel events

azgron committed:
Pass user settings to server

m4ntis committed:
pf_server: Proxy mouse events

kubistika committed:
fixup! server/proxy: Add graphics callbacks

m4ntis committed:
pf_client: Fix setting initialization

m4ntis committed:
Merge feat/proxy-gfx to feat/proxy

m4ntis committed:
pf_server: Fix double freed credentials

kubistika committed:
server/proxy: Remove unnecessary call to freerdp_client_settings_parse_command_line

kubistika committed:
server/proxy: Refactor re-activation code

kubistika committed:
server/proxy: Run format scripts

kubistika committed:
server/proxy: Fix segfault when post_disconnect return FALSE

kubistika committed:
server/proxy: Refactor proxy_settings_mirror

kubistika committed:
server/proxy: Redirect credentials

kubistika committed:
server/proxy: move proxy_settings_mirror to pf_common.c

kubistika committed:
server/proxy: Redirect desktop_resize event

m4ntis committed:
pf_client: Remove interactive CLI auth methods

kubistika committed:
fixup! server/proxy: Redirect credentials

kubistika committed:
server/proxy: Rename proxy_mirror_settings to pf_common_copy_settings

kubistika committed:
pf_server.c: Fixed non-freed context

When the disconnection is forced by the target server, the function
`pf_server_handle_client_disconnection` isn't called. Therefore, the
context of the connection between the proxy to tagrget isn't freed.

kubistika committed:
fixup! pf_server.c: Fixed non-freed context

m4ntis committed:
pf_client: Prefix all client methods with pf_client

m4ntis committed:
pf_context: Add init client to proxy context method

m4ntis committed:
pf_server: Confirm all GFX caps regardless of settings

m4ntis committed:
pf_server: Prefix all methods with pf_server

m4ntis committed:
pf_server: Move variable decleration to start of method

m4ntis committed:
pf_server: Fix client setting

m4ntis committed:
pf_server: Fix GFX init method

m4ntis committed:
pf_server: Move variable decleration to start of methods

m4ntis committed:
server/proxy: Formatting

m4ntis committed:
Merge feat/proxy

m4ntis committed:
pf_server: Proxy synchronize event

m4ntis committed:
pf_server: Proxy refresh rect update events

m4ntis committed:
pf_server: Proxy suppress output messages

m4ntis committed:
server/proxy: Fix licensing

m4ntis committed:
server/proxy: Move client input callbacks to pf_input

m4ntis committed:
server/proxy: Move client update callbacks to pf_update

akallabeth committed:
Added python generator for freerdp_[g|s]et_param*

The settings getters and setters are now automatically generated.
Added a test case to check if all available options can be retrieved.

kubistika committed:
server/proxy: Fix non-terminated target host string

akallabeth committed:
Fixed index data type

akallabeth committed:
Fixed possible memory leak.

akallabeth committed:
Fixed memory leak

akallabeth committed:
Fixed memory leak

akallabeth committed:
Fixed memory leak

akallabeth committed:
Fixed uninitialized value

akallabeth committed:
For smartcard trace functions use const arguments

akallabeth committed:
Unified update->BeginPaint and update->EndPaint

Since these functions were called from 2 different threads
(main thread or dynamic channel) depending on fastpath or
gfx channel use lock between these.
If not locked the invalid region may be accessed from both
threads and lead to crashes as experienced with nightly df280a7ff

kubistika committed:
Feat/proxy config (#2)

* server/proxy: Add config loading support

* server/proxy: Add config file

* server/proxy: Format code

* server/proxy: Code refactor, rename update_register_callbacks and input_register_callbacks

* server/proxy: Update config file

* server/proxy: Remove config.ini from root directory

* Remove comment from config file

* server/proxy: Fix leak in pf_server_load_config

* server/proxy: Add rdpServerProxy struct and embed it in proxyContext

* server/proxy: Load configuration and pass it inside every proxyContext instance

* server/proxy: Move rdpProxyServer to proxy.h

* server/proxy: Use configuration while proxying input events

* server/proxy: Update CMakeLists

* server/proxy: Refactor pf_input.c

* server/proxy: Add AllowedChannels, DeniedChannels in configuration

* server/proxy: Remove unnecessary variable from parse_channels_from_str

* server/proxy: Update config file

* server/proxy: config: Rename  to

* server/proxy: config: Add mode - blacklist/whitelist

* server/proxy: Refactor, fix NULL deref

* server/proxy: Add license to proxy.h

* server/proxy: Fix newline in pf_config.c

* server/proxy: config: Rename Mode to WhitelistMode

kubistika committed:
Add target in config. Add checks for configuration validity (#3)

* Add target in config. Add checks for configuration validity

* Update config file

speidy committed:
libfreerdp: nego: revert commented out check of routingToken length

kubistika committed:
pf_server: Fix target host info from RoutingToken

kubistika committed:
pf_server: Remove hardcoded lenght of routing token prefix

kubistika committed:
Feat/refactor context (#8)

* Refactor main structs

* Update CMakeLists.txt

* pf_server.c: Free pdata at the end of the connection

* Run format scripts

* Rename tf to pc

* Fix licenses

* pf_server: Refactor names of structs and functions

speidy committed:
proxy: gfx: sync caps (#4)

* proxy: gfx: sync caps

* proxy: gfx: sync caps, hook gfx client's OnClose() call and close server resources

speidy committed:
fixup! Feat/refactor context (#8)

kubistika committed:
fixup! fixup! Feat/refactor context (#8)

kubistika committed:
rdpgfx/client: Fix rdpgfx_recv_caps_confirm_pdu caps set length parsing

kubistika committed:
Run format scripts

m4ntis committed:
proxy config.ini: Change default port to 3389

m4ntis committed:
pf_rdpgfx: Limit caps version to freerdp's supported versions

kubistika committed:
rdpgfx: Fix GFX v10.6 PDUs parsing and naming according to the spec

kubistika committed:
pf_rdpgfx: Proxy rdpgfx v10.6 PDUs

m4ntis committed:
Gfx OnOpen() wait for dynvc ready (#10)

* proxy/gfx: Wait for dynvc ready state before open

* pf_channels: Initialize pc->gfx

* pf_rdpgfx: Add log and fix comments

m4ntis committed:
gfx client: Publish FrameAck sending and add auto ack flag

m4ntis committed:
proxy/gfx: Forward frame ack messages

m4ntis committed:
pf_context: Forward domain on connection

m4ntis committed:
pf_rdpgfx: Change max supported caps to 10.6

speidy commented on include/freerdp/server/rdpgfx.h:
```diff
@@ -21,6 +21,7 @@
 #define FREERDP_CHANNEL_RDPGFX_SERVER_RDPGFX_H
 
 #include <freerdp/channels/rdpgfx.h>
+#include <freerdp/freerdp.h>
```
shouldn't be here , but in pf_server.c i think

m4ntis replied:
test it

speidy commented on server/proxy/CMakeLists.txt:
```diff
@@ -0,0 +1,76 @@
+  pf_graphics.h
+  pf_log.h)
+
+	# On windows create dll version information.
```
take me back

kubistika replied:
done

speidy commented on server/proxy/pf_channels.c:
```diff
@@ -0,0 +1,137 @@
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT pf_encomsp_participant_created(EncomspClientContext* context,
```
maybe remove `encomsp` stuff for now

m4ntis replied:
oh yeah for sure, there is the encomsp in pf_context.h too if i remember correctly, remove it all

speidy commented on server/proxy/pf_channels.c:
```diff
@@ -0,0 +1,137 @@
+	{
+		pc->rdpei = (RdpeiClientContext*) e->pInterface;
+	}
+	else if (strcmp(e->name, TSMF_DVC_CHANNEL_NAME) == 0)
```
rm unhandled channels ?

m4ntis replied:
yes

speidy commented on server/proxy/pf_channels.c:
```diff
@@ -0,0 +1,137 @@
+	}
+	else if (strcmp(e->name, ENCOMSP_SVC_CHANNEL_NAME) == 0)
+	{
+		pf_encomsp_init(pc, (EncomspClientContext*) e->pInterface);
```
whats this `encomsp` ?? :|

m4ntis replied:
i answered above

speidy commented on server/proxy/pf_client.c:
```diff
@@ -0,0 +1,421 @@
+void proxy_server_reactivate(rdpContext* client, rdpContext* target)
+{
+	pf_common_copy_settings(client->settings, target->settings);
+	client->update->DesktopResize(client);
```
@kubistika  add comment why we call this func

speidy commented on server/proxy/pf_client.c:
```diff
@@ -0,0 +1,421 @@
+	update->BitmapUpdate = pf_client_bitmap_update;
+	update->DesktopResize = pf_client_desktop_resize;
+	rdpContext* ps = (rdpContext*) pc->pdata->ps;
+	proxy_server_reactivate(ps, context);
```
should we call this here?

kubistika replied:
What do you mean by this? The reactivation? Yes, we reactivate right after the connection

speidy commented on server/proxy/pf_client.c:
```diff
@@ -0,0 +1,421 @@
+}
+
+
+static void pf_client_client_free(freerdp* instance, rdpContext* context) {}
```
nothing to do here?

speidy commented on server/proxy/pf_common.c:
```diff
@@ -0,0 +1,106 @@
+	dst->GfxAVC444 = src->GfxAVC444;
+	dst->GfxSendQoeAck = src->GfxSendQoeAck;
+	dst->GfxAVC444v2 = src->GfxAVC444v2;
+}
```
new line @ eof

speidy commented on server/proxy/pf_context.c:
```diff
@@ -0,0 +1,80 @@
+		goto fail_open_server;
+
+	return TRUE;
+fail_open_server:
```
add new line before label

speidy commented on server/proxy/pf_rdpgfx.c:
```diff
@@ -0,0 +1,306 @@
+	gfx->EvictCacheEntry = pf_rdpgfx_evict_cache_entry;
+	gfx->MapSurfaceToOutput = pf_rdpgfx_map_surface_to_output;
+	gfx->MapSurfaceToWindow = pf_rdpgfx_map_surface_to_window;
+    gfx->MapSurfaceToScaledOutput = pf_rdpgfx_map_surface_to_scaled_output;
```
spaces to tabs

speidy commented on server/proxy/pf_server.c:
```diff
@@ -0,0 +1,432 @@
+
+#define TAG PROXY_TAG("server")
+
+BOOL pf_server_rdpgfx_init(pServerContext* ps)
```
move to `pf_rdpgfx.c`

speidy commented on server/proxy/pf_server.c:
```diff
@@ -0,0 +1,432 @@
+	{
+		/* use hardcoded target info from configuration */
+		host = _strdup(config->TargetHost);
+		port = config->TargetPort > 0 ? config->TargetPort : port;
```
`&& config->TargetPort <= USHORT_MAX`

speidy commented on server/proxy/pf_server.c:
```diff
@@ -0,0 +1,432 @@
+	/* Start a proxy's client in it's own thread */
+	pc = (pClientContext*) p_client_context_create(client->settings, host, port);
+	pc->pdata = ps->pdata;
+	/* Inject proxy's client context to proxy's context */
```
Inject -> keep

speidy commented on server/proxy/pf_server.c:
```diff
@@ -0,0 +1,432 @@
+	ps->dynvcReady = CreateEvent(NULL, TRUE, FALSE, NULL);
+	pdata = calloc(1, sizeof(proxyData));
+	ps->pdata = pdata;
+	/* inject configuration to proxyData */
```
again , not inject

speidy commented on server/proxy/pf_server.c:
```diff
@@ -0,0 +1,432 @@
+	client->settings->SupportGraphicsPipeline = TRUE;
+	client->settings->SupportDynamicChannels = TRUE;
+	/* TODO: Read path from config and default to /etc */
+	client->settings->CertificateFile = _strdup("server.crt");
```
lets add to `config.ini`, its easy

m4ntis replied:
do it

speidy commented on server/proxy/pf_server.c:
```diff
@@ -0,0 +1,432 @@
+	/* TODO: Read path from config and default to /etc */
+	client->settings->CertificateFile = _strdup("server.crt");
+	client->settings->PrivateKeyFile = _strdup("server.key");
+	client->settings->RdpKeyFile = _strdup("server.key");
```
rm, its not relevant

m4ntis replied:
do it

kubistika replied:
done

speidy commented on server/proxy/pf_server.c:
```diff
@@ -0,0 +1,432 @@
+	if (!client->settings->CertificateFile || !client->settings->PrivateKeyFile
+	    || !client->settings->RdpKeyFile)
+	{
+		WLog_ERR(TAG, "Memory allocation failed (strdup)");
```
change the log so it mentions `CertificateFile`  `PrivateKeyFile` etc..

m4ntis replied:
go for it

m4ntis commented on include/freerdp/server/rdpgfx.h:
```diff
@@ -21,6 +21,7 @@
 #define FREERDP_CHANNEL_RDPGFX_SERVER_RDPGFX_H
 
 #include <freerdp/channels/rdpgfx.h>
+#include <freerdp/freerdp.h>
```
Need to ensure this is legit

speidy replied:
it wasn't there, and doesn't make sense

m4ntis replied:
I'm not sure it compiles without it tho, can you please test that?

speidy commented on server/proxy/pf_server.c:
```diff
@@ -0,0 +1,432 @@
+	if (!client->settings->CertificateFile || !client->settings->PrivateKeyFile
+	    || !client->settings->RdpKeyFile)
+	{
+		WLog_ERR(TAG, "Memory allocation failed (strdup)");
```
change the log so it mentions `CertificateFile`  `PrivateKeyFile` etc..

m4ntis commented on libfreerdp/core/nego.c:
```diff
@@ -36,34 +36,6 @@
 
 #define TAG FREERDP_TAG("core.nego")
 
-struct rdp_nego
```
@kubistika I can't remember why we did this

speidy replied:
i did that so we can get the routing token from our code.

m4ntis replied:
So can we revert it

kubistika replied:
No, we use rdpNego struct in pf_server to get the routing token and extract target host information. This struct should remain exported.

speidy commented on server/proxy/pf_server.c:
```diff
@@ -0,0 +1,432 @@
+	}
+
+fail:
+
```
rm blank line

m4ntis replied:
yes

kubistika replied:
but the formatter adds it

m4ntis committed:
proxy: Update config

speidy reviewed (changes requested):
Take a look at my comments, lets merge

speidy commented:
we should squash the `fixup!` commits using `git rebase -i --autosquash origin/master`

kubistika committed:
server/proxy: Use configuration in pf_server_handle_client

kubistika committed:
rdpgfx/client: Fix size of surface_to_scaled_window, surface_to_window

kubistika committed:
pf_rdpgfx: Fix formatting

kubistika committed:
pf_server.c: Fix comments

kubistika committed:
Move pf_server_rdpgfx_init to pf_rdpgfx

kubistika committed:
server/proxy/CMakeLists.txt: Fix formatting

kubistika committed:
pf_client.c: Add comment in proxy_server_reactivate

Commit:
Fixed const correctness of gfx function pointer

Signed-off-by: Mati Shabtay <matishabtay@gmail.com>

speidy committed:
server: proxy: update copyright

speidy committed:
server: proxy: wrap rdpNego and add a getter for routing token

kubistika committed:
Refactor routing token getter (#14)

* Refactor routing token getter

* pf_server_parse_target_from_routing_token change routing_token_length to be DWORD

* libfreerdp/core/nego.c: Run format script

kubistika committed:
pf_server: Run format script

m4ntis committed:
server/proxy: Fix os msbuild tests

m4ntis committed:
pf_channels.c: Remove unused channels

m4ntis committed:
pf_client: Remove unused callbacks

m4ntis committed:
proxy: Remove encomsp callbacks from proxy's client

m4ntis committed:
client/rdpgfx_main.c: Fix msbuild test

m4ntis committed:
pf_config.c: Use StrSep instead of strsep for Windows builds

Commit:
Removed nego struct from direct access.

Signed-off-by: Mati Shabtay <matishabtay@gmail.com>

kubistika committed:
rdpgfx: Fix GFX v10.6 PDUs parsing and naming according to the spec

m4ntis committed:
client/rdgpfx_main.c: Fix rdpgfx_recv_caps_confirm_pdu caps length read

m4ntis committed:
proxy: Rename binary to freerdp-proxy

akallabeth committed:
Fixed use of settings getter/setter

akallabeth committed:
Merge pull request #5377 from m4ntis/gfx/fix-caps-confirm-recv

client/rdgpfx_main.c: Fix rdpgfx_recv_caps_confirm_pdu caps length read

akallabeth committed:
Added image scaling api for software drawing.

For future GFX channel functions an image scaling function is required.
This moves the implementation from wayland client to core library
and adds support for the much faster SWScale library.

akallabeth committed:
Merge pull request #5376 from kubistika/bugfix/rdpgfx-server-v10.6-PDUs

rdpgfx: Fix GFX v10.6 PDUs parsing and naming according to the spec

m4ntis committed:
rdpgfx_main.c: Revert unwanted double change to send_supported_caps

akallabeth committed:
Updated GFX to 10.6 spec

mfleisz committed:
Merge pull request #5381 from akallabeth/swscale

Added image scaling api for software drawing.

akallabeth committed:
Optimized image scaling, copy if factor is 1.0

akallabeth committed:
Fixed missing argument checks.

The arguments for rdpgfx_send_frame_acknowledge_pdu must be checked
as in some call stacks they might be NULL.

m4ntis committed:
Add .md file extention to README

mfleisz committed:
Merge pull request #5339 from akallabeth/update_paint_lock

Unified update->BeginPaint and update->EndPaint

hardening committed:
Merge pull request #5384 from m4ntis/fix-readme

Add .md file extention to README

akallabeth committed:
Fixed alignment requirements for surface sizes.

akallabeth committed:
Added gfx capability filter option.

With this new option it is possible to selectively disable certain
GFX capability versions. This way buggy behaviour can be addressed
at runtime.

akallabeth committed:
Fixed memory leak and return values in rdpgfx_recv_caps_advertise_pdu

akallabeth committed:
Silenced some unused parameter warnings.

akallabeth committed:
Silenced unused parameter warnings in smartcard trace

akallabeth committed:
Cleaned up proxy server code.

* All internal functions static
* Added simple command line argument to supply a config file
* Silence compiler warnings

Signed-off-by: kubistika <kmizrachi18@gmail.com>

akallabeth committed:
Fixed doulbe semicolon in C files.

mfleisz committed:
Merge pull request #5383 from akallabeth/gfx_spec_update

Updated GFX to 10.6 spec

mfleisz committed:
Merge pull request #5385 from akallabeth/some_unused_parameter_silenced

Some unused parameter silenced

akallabeth committed:
Refactored settings getter/setter to stay compatible.

akallabeth committed:
Fixed error checks and uninitialized values in convert_filedescriptors_to_uri_list

mfleisz committed:
core: Move image conversion for icon data into the library

mfleisz committed:
Merge pull request #5266 from akallabeth/file_list_synth

Added a proper synthesizer for FileGroupDescriptorW to text/uri-list

akallabeth committed:
Fixed warnings and cleaned up gfx related xfreerdp code.

akallabeth committed:
Updated RAILS implementation

* Implement new messages and callbacks
* Announce most recent channel features
* Added settings to configure flags to announce

akallabeth committed:
Merge pull request #5387 from mfleisz/image_from_icon

core: Move image conversion of icon data into the library

akallabeth committed:
Enable cairo support for software scaling.

hardening committed:
Merge pull request #5388 from akallabeth/nightly_cairo

Nightly: activate cairo software scaling

akallabeth committed:
Fixed rails set_caretwidth

akallabeth committed:
Fixed GFX flags and CapsConfirm return checks

mfleisz committed:
Merge pull request #5386 from akallabeth/rails_spec_update

Rails spec update

mfleisz committed:
Merge pull request #5392 from akallabeth/gfx_avc_flag_fixes

Fixed GFX flags and CapsConfirm return checks

mfleisz committed:
Merge pull request #5327 from akallabeth/sign_compare_fixes_v3

Leak and warning fixes for 2.0

akallabeth committed:
Fixed window order cleanup for NOTIFY_ICON_STATE_ORDER

mfleisz committed:
Merge pull request #5393 from akallabeth/rails_memleak_fix

Fixed window order cleanup for NOTIFY_ICON_STATE_ORDER

akallabeth committed:
Fixed remaining getter/setter return checks.

mfleisz committed:
Merge pull request #5314 from akallabeth/setting_getter_setter_generator

Added python generator for freerdp_[g|s]et_param*

akallabeth committed:
Updated settings getter/setter for RemoteApplicationSupportMask

mfleisz committed:
Merge pull request #5394 from akallabeth/settings_update

Updated settings getter/setter for RemoteApplicationSupportMask

speidy commented:
lets close this one

kubistika committed:
bugfix: segfault when calling an event which is not implemented by filter

kubistika committed:
Add README for filters API

kubistika committed:
server/proxy: Add external filters support

kubistika committed:
server/proxy: Refactor pf_server_parse_target_from_routing_token

kubistika committed:
server/proxy: Refactor pf_server.c

kubistika committed:
server/proxy: Use new parse_string_array_from_str implementation (using ArrayList)

kubistika committed:
server/proxy: Add return statement in demo_filter.c

kubistika committed:
server/proxy: Add docs in filters_api.h

kubistika committed:
server/proxy: Remove unnecessary wrap for connection info

speidy committed:
Merge pull request #15 from m4ntis/feat/proxy-filters

server/proxy: Add external filters support

kubistika committed:
server/proxy: Fix free invalid memory when pf_server_parse_target_from_routing_token fails

speidy committed:
Merge pull request #16 from m4ntis/bugfix/segfault-free-invalid-mem

server/proxy: Fix free invalid memory
